# 검정(Test)
---

## Z-검정(Z-Test)

$$Z = \frac{\bar{X} - \mu}{\frac{\sigma}{\sqrt{n}}}$$
$$\bar{X} : 표본평균 \\
\mu : 모평균 \\
\sigma : 모표준편차 \\
n : \text{표본의 크기}$$

- 공장에서 생성한 제품의 평균 높이는 26cm이며 표준편차는 5cm로 확인되었다. 제품의 높이를 측정한 표본 7개의 데이터는 다음과 같다. 유의수준이 `0.05`일 때, 데이터를 기반으로 제품의 평균 높이를 검정하려고 한다.

| 제품의 높이(cm) | 25, 27, 31, 23, 24, 30, 26 |
| --- | --- |
| 귀무가설($𝐻_0$) | 모평균은 26cm (모평균 = 표본평균) |
| 대립가설($𝐻_1$) | 모평균은 26cm가 아님 (모평균 ≠ 표본평균) |

In [8]:
import numpy as np
from scipy.stats import norm

x = np.array([25, 27, 31, 23, 24, 30, 26])
print(f"mean: {np.mean(x)}")   # 평균 출력

mean: 26.571428571428573


$$Z = \frac{X-26}{\frac{5}{\sqrt{7}}} = \frac{(X-26)\sqrt{7}}{5}$$

In [9]:
# z 값 계산
## 모평균 : 26
## 표준편차 : 5
## 표본 개수 : 7
z = (np.mean(x) - 26) / (5 / np.sqrt(7))
print(f"z: {z}")

z: 0.3023715784073826


In [10]:
# p-값 (유의수준) 출력
p = (1 - norm.cdf(z)) * 2    # z의 위치까지 누적 분포 함수의 값 계산 후, p-값을 얻기 위해 전체 면적 1에서 뺌. 양측 검정이므로 2배를 곱함.
print(f"p: {p}")

p: 0.7623688184698392


## T-검정(T-Test)

$$T = \frac{\bar{X} - \mu}{\frac{s}{\sqrt{n}}}$$
$$\bar{X} : 표본평균 \\
\mu : 모평균 \\
s: 표본표준편차 \\
n : \text{표본의 크기}$$

### 단일 표본 T-검정(One Sample T-Test)

#### 정규성 가정을 만족하는 경우 단일 표본 T-검정

In [14]:
import pandas as pd
from scipy.stats import shapiro, ttest_1samp

df = pd.DataFrame({'height': [12, 14, 16, 19, 11, 17, 13]})
print(df)

   height
0      12
1      14
2      16
3      19
4      11
5      17
6      13


- 데이터가 정규 분포를 따르는지 확인하기 위해 **샤피로 윌크 검정**을 수행한다.

In [15]:
print(shapiro(df['height']))

ShapiroResult(statistic=0.9641614571212591, pvalue=0.8535413398236595)


- p값은 `0.8535`이므로 유의수준 `0.05`보다 <ins>크기</ins> 때문에 귀무가설(정규성 만족)을 채택한다.
- `df`의 `height` 변수는 정규성을 만족하므로 `ttest_1samp` 함수를 사용하여 검정을 수행한다.

In [16]:
print(ttest_1samp(df['height'], popmean=11))

TtestResult(statistic=3.2826608214930637, pvalue=0.016766749930606027, df=6)


- p-값은 `0.0168`로 유의수준 `0.05`보다 <ins>작기</ins> 때문에 귀무가설을 기각하고 대립가설을 채택한다.

#### 정규성 가정을 만족하지 않는 경우 단일 표본 T-검정

- `cats` 데이터는 고양이들의 성별, 몸무게, 심장의 무게를 담고 있다. 유의수준 `0.05`일 때 `cats` 데이터에서 고양이들의 평균 몸무게가 2.1kg인지 아닌지에 대해 **양측 검정**을 수행한다.

| 귀무가설($𝐻_0$) | 고양이들의 평균 몸무게(Bwt 변수 값의 평균)는 2.1kg |
| --- | --- |
| 대립가설($𝐻_1$) | 고양이들의 평균 몸무게(Bwt 변수 값의 평균)는 2.1kg가 아님. |

In [11]:
import pandas as pd
from scipy.stats import shapiro, wilcoxon

df = pd.read_csv('./datasets/cats.csv')
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144 entries, 0 to 143
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   rownames  144 non-null    int64  
 1   Sex       144 non-null    object 
 2   Bwt       144 non-null    float64
 3   Hwt       144 non-null    float64
dtypes: float64(2), int64(1), object(1)
memory usage: 4.6+ KB
None


- 데이터가 정규 분포를 따르는지 확인하기 위해 **샤피로 윌크 검정**을 수행한다.

In [12]:
result = shapiro(df['Bwt'])   # 샤피로 검정 수행
print(result)

ShapiroResult(statistic=0.9518791269479144, pvalue=6.730857622701013e-05)


- p값은 `6.730×10^-5`이므로 유의수준 `0.05`보다 <ins>작기</ins> 때문에 귀무가설을 기각하고 대립가설(정규성 불만족)을 채택한다.
- `df`의 `height` 변수는 정규성을 불만족하므로 윌콕슨 검정을 검정을 수행한다.

In [13]:
result = wilcoxon(df['Bwt'] - 2.1, alternative='two-sided')   # 평균 몸무게가 2.1kg인지 아닌지에 대해 양측 검정 수행
print(result)

WilcoxonResult(statistic=50.0, pvalue=2.765612175340855e-23)


- 윌콕슨 검정 통계량은 `50.0`이고, p-값은 유의 수준 `0.05`보다 <ins>작기</ins> 때문에 귀무가설을 기각하고 대립가설을 채택한다.

### 쌍체 표본 T-검정(Paired Sample T-Test; 대응 표본 T-검정)

$$T = \frac{\bar{d} - \mu_{d}}{\frac{s}{\sqrt{n}}}$$
$$\bar{d} : \text{두 표본 집단 평균의 차이} \\
\mu_{d} : \text{두 모집단 평균의 차이} \\
s: 표본표준편차 \\
n : \text{표본의 크기}$$

In [17]:
import pandas as pd
from scipy.stats import ttest_rel

data = pd.DataFrame({
    'before': [5, 3, 8, 4, 3, 2, 1],
    'after': [8, 6, 6, 5, 8, 7, 3]
})

print(data)

   before  after
0       5      8
1       3      6
2       8      6
3       4      5
4       3      8
5       2      7
6       1      3


In [18]:
# 수면 영양제를 복용하기 전과 후의 평균 수명 시간 차이가 비교하고자 하는 값(0) 보다 작은지에 대해 검정을 수행하기 때문에 less 입력
result = ttest_rel(data['before'], data['after'], alternative='less')
print(result)

TtestResult(statistic=-2.633628675421043, pvalue=0.019435182851729293, df=6)


In [19]:
print(result.pvalue)   # p-값 출력

0.019435182851729293


- 유의수준 `0.05`일 경우, p-값은 `0.01944`로 유의수준보다 작기 때문에 귀무가설을 기각하고 대립가설을 채택한다.

### 독립 표본 T-검정(Independent Sample T-Test)

- `cats` 데이터는 고양이들의 성별, 몸무게, 심장의 무게를 담고 있다. 유의수준 `0.05`일 때 고양이들의 성별에 따른 몸무게의 평균은 통계적으로 다르다고 할 수 있는지에 대한 검정을 수행하려고 한다.

| 귀무가설($𝐻_0$) | 고양이의 성별에 따른 평균 몸무게는 통계적으로 유의한 차이가 없음. |
| --- | --- |
| 대립가설($𝐻_1$) | 고양이의 성별에 따른 평균 몸무게는 통계적으로 유의한 차이가 있음.  |

In [20]:
import pandas as pd
from scipy.stats import ttest_ind, levene

cats = pd.read_csv('./datasets/cats.csv')
print(cats.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144 entries, 0 to 143
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   rownames  144 non-null    int64  
 1   Sex       144 non-null    object 
 2   Bwt       144 non-null    float64
 3   Hwt       144 non-null    float64
dtypes: float64(2), int64(1), object(1)
memory usage: 4.6+ KB
None


In [21]:
group1 = cats[cats['Sex'] == 'F']['Bwt']
group2 = cats[cats['Sex'] == 'M']['Bwt']

result = levene(group1, group2)
print(result)

LeveneResult(statistic=19.43101190877999, pvalue=2.0435285255189404e-05)


In [22]:
print(result.pvalue)

2.0435285255189404e-05


- p-value가 유의 수준보다 작기 때문에 귀무가설(등분산성 가정 만족) 기각하고, 대립가설(등분산성 가정 불만족) 채택


In [23]:
result = ttest_ind(group1, group2, equal_var=False)
print(result)

TtestResult(statistic=-8.70948849909559, pvalue=8.831034455859356e-15, df=136.83788299625363)


In [24]:
print(result.pvalue)

8.831034455859356e-15


- p-값은 $8.831×10^{-15}$로 유의수준(0.05)보다 작기 때문에 귀무가설을 기각하고, 대립가설을 채택한다.

## F-검정(F-Test)

In [39]:
import numpy as np
from scipy.stats import f

df1 = np.array([1, 2, 3, 4, 6])
print(np.var(df1, ddof=1))   # 분산 출력 (자유도: 1)

3.7


In [40]:
df2 = np.array([4, 5, 6, 7, 8])
print(np.var(df2, ddof=1))

2.5


In [41]:
def f_test(x, y):
    if np.var(x, ddof=1) < np.var(y, ddof=1):
        x, y = y, x
    
    f_value = np.var(x, ddof=1) / np.var(y, ddof=1)
    x_dof = x.size - 1
    y_dof = y.size - 1
    p_value = (1 - f.cdf(f_value, x_dof, y_dof)) * 2   # 양측 검정
    
    return f_value, p_value

In [42]:
result = f_test(df1, df2)
print(result)

(1.48, 0.7133026753046221)


## 카이제곱 검정(Chi-Squared Test; $\chi^{2}$ Test)

### 적합도 검정(Goodness of Fit Test)

- 유의수준이 `0.05`일 때 초등학교에 남학생이 90명, 여학생이 160명이 있다. 남학생, 여학생 비율이 45%와 55%인지를 카이제곱 검정을 이용하여 분석하려고 한다.

| 귀무가설($𝐻_0$) | 초등학교 남학생, 여학생 비율은 45%와 55% |
| --- | --- |
| 대립가설($𝐻_1$) | 초등학교 남학생, 여학생 비율이 45%와 55%가 아님. |

In [2]:
import numpy as np
from scipy.stats import chisquare

num = np.array([90, 160])

# 기대 빈도 계산
expected = np.array([0.45, 0.55]) * np.sum(num)

result = chisquare(num, f_exp=expected)
print(result)

Power_divergenceResult(statistic=8.181818181818182, pvalue=0.004231232899758152)


In [3]:
print(result.pvalue)

0.004231232899758152


### 독립성 검정(Test of Independence)

In [6]:
import pandas as pd
from scipy.stats import chi2_contingency

survey = pd.read_csv('./datasets/survey.csv')

tb = pd.crosstab(survey['Sex'], survey['Exer'])   # 각 변수의 빈도수를 계산하여 변수에 저장
print(tb)

Exer    Freq  Some
Sex               
Female    49    58
Male      65    40


In [7]:
result = chi2_contingency(tb)
print(result)

Chi2ContingencyResult(statistic=4.904232352768243, pvalue=0.0267909570897706, dof=1, expected_freq=array([[57.53773585, 49.46226415],
       [56.46226415, 48.53773585]]))


In [8]:
print(result.pvalue)

0.0267909570897706


### 동질성 검정(Test of Homogeneity)

In [9]:
import numpy as np
from scipy.stats import chi2_contingency

data = np.array([[20, 30, 50], [30, 50, 20]])

chi2, p, dof, expected = chi2_contingency(data)

print(f"Chi-squared value: {chi2}")
print(f"p-value: {p}")
print(f"Degrees of freedom: {dof}")
print("Expected frequencies:")
print(expected)

alpha = 0.05

if p < alpha:
    print("귀무가설 기각: 두 모집단은 동일한 분포를 갖지 않습니다.")
else:
    print("귀무가설 채택: 두 모집단은 동일한 분포를 가집니다.")

Chi-squared value: 19.857142857142858
p-value: 4.8761405516518554e-05
Degrees of freedom: 2
Expected frequencies:
[[25. 40. 35.]
 [25. 40. 35.]]
귀무가설 기각: 두 모집단은 동일한 분포를 갖지 않습니다.
